### PR TITLE
Balanced datasets

### PR DESCRIPTION
**Pull request by ngreenwald**
This PR adds the functionality to balance datasets so that different classes are represented equally. 
This is accomplished by identifying the category with the largest number of images, and then randomly sampling with replacement from the other categories to achieve the same number of total images. 

When the balance_dataset flag is set to true, this is performed for the train and val splits, but not for the test split. The rationale is that we want to be able to assess how balanced training effects the accuracy on the test dataset, and if that was upsampled, we wouldn't be able to compare the effects. A balanced F1 score can be calculated via weighting the results. 

In addition, I switched the representation of the metadata from lists to numpy arrays to avoid the constant swapping back and forth, since there weren't actually any benefits from keeping it as a list format.

### CODE REVIEW

File: caliban_toolbox/dataset_builder.py [hunk 1]
```diff
@@ -508,7 +564,7 @@ def _validate_output_shape(self, output_shape):
                              'or length 3, got {}'.format(output_shape))
 
     def build_dataset(self, tissues='all', platforms='all', output_shape=(512, 512), resize=False,
-                      data_split=(0.8, 0.1, 0.1), seed=0, **kwargs):
+                      data_split=(0.8, 0.1, 0.1), seed=0, balance_dataset=False, **kwargs):
```
Comment (willgraf) [inline caliban_toolbox/dataset_builder.py:567]: Can we shorten this to just `balance`?

File: caliban_toolbox/dataset_builder.py [hunk 2]
```diff
@@ -534,7 +592,7 @@ def build_dataset(self, tissues='all', platforms='all', output_shape=(512, 512),
         Raises:
             ValueError: If invalid resize parameter supplied
         """
-        if self.all_tissues == []:
+        if self.all_tissues is None:
```
Comment (willgraf) [inline caliban_toolbox/dataset_builder.py:595]: You could generalize this to `if not self.all_tissues` which would work for `None` and empty lists/arrays.

**Review by willgraf — CHANGES_REQUESTED** [review]
Looks great! I really like the list-> np.array refactor, much cleaner. I just have a couple small suggestions.

File: caliban_toolbox/dataset_builder.py (hunk 2 above)
Comment (ngreenwald) [reply caliban_toolbox/dataset_builder.py:595]: Just tried this, the problem is that when it's not an empty array, leads to the following error:
`ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
`

**Review by willgraf — APPROVED** [review]
